lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -674,6 +730,83 @@ public void OnNext(IFailedTask failedTask)
         }
         #endregion IFailedTask
 
+        private void TimeoutMonitor(object source, ElapsedEventArgs e)
+        {
+            Logger.Log(Level.Info, "Entering TimeoutMonitor at {0}", DateTime.Now);
+            lock (_lock)
+            {
+                switch (_systemState.CurrentState)
+                {
+                    // TODO: Handle time out if ActiveContexts are not received in timeout limit
+                    case SystemState.WaitingForEvaluator:
+                        break;
+
+                    // TODO: Handle time out if RunningTasks are not received in timeout limit
+                    case SystemState.SubmittingTasks:
+                        break;
+
+                    // TODO: Handle time out if CompletedTasks are not received in timeout limit
+                    case SystemState.TasksRunning:
+                        break;
+
+                    // Handle timeout for closing tasks
+                    case SystemState.ShuttingDown:
+                        Logger.Log(Level.Info, "_taskManager.AverageClosingTime {0}, _minTaskWaitingForCloseTimeout: {1}", _taskManager.AverageClosingTime(), _minTaskWaitingForCloseTimeout);
+                        int taskClosingTimeout = Math.Max(_minTaskWaitingForCloseTimeout, _taskManager.AverageClosingTime() * TaskWaitingForCloseTimeFactor);
+                        var waitingTasks = _taskManager.TasksWaitingForClose(taskClosingTimeout);
+
+                        if (waitingTasks.Count > 0)
+                        {
+                            WaitingForCloseTaskNoResponseAction(waitingTasks);
+                        }
+                        break;
+
+                    case SystemState.TasksCompleted:
+                        break;
+
+                    case SystemState.Fail:
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// For tasks that are in WaitingForCloseState and has no response in specified timeout
+        /// kill the evaluator and set the other states as if we received the FailedEvaluator
+        /// Then try recovery
+        /// </summary>
+        /// <param name="tasks"></param>
+        private void WaitingForCloseTaskNoResponseAction(IList<KeyValuePair<string, TaskInfo>> tasks)
+        {
+            foreach (var t in tasks)
+            {
+                string evaluatorId = t.Value.ActiveContext.EvaluatorId;
+                if (!_evaluatorsForceClosed.Contains(evaluatorId))
+                {
+                    _evaluatorsForceClosed.Add(evaluatorId);
+                    Logger.Log(Level.Info,
+                        "WaitingForCloseTask [{0}] has no response after [{1}]. Kill the evaluator: [{2}] and context: [{3}].",
+                        t.Key,
+                        _minTaskWaitingForCloseTimeout,
```

Comment:
We're logging this value but using `taskClosingTimeout` (which is greater) to actually determine when to kill the tasks. I'd remove this value from this log to avoid confusion